ostd/src/arch/x86/timer/apic.rs

```diff
@@ -28,49 +20,79 @@ use crate::{
 };
 
+pub(super) static IN_APIC_TSC_DEADLINE_MODE: AtomicBool = AtomicBool::new(false);
+
 /// Initializes APIC with tsc deadline mode or periodic mode.
-/// Return the corresponding [`IrqLine`] for the System Timer.
-pub(super) fn init() -> IrqLine {
+/// Return the corresponding [`IrqLine`] for the system timer.
+pub(super) fn init_bsp() -> IrqLine {
     init_tsc_freq();
+
     if is_tsc_deadline_mode_supported() {
         info!("[Timer]: Enable APIC TSC deadline mode.");
-        init_tsc_mode()
+        IN_APIC_TSC_DEADLINE_MODE.store(true, Ordering::Relaxed);
+
+        let timer_irq = IrqLine::alloc().unwrap();
+        init_tsc_mode(&timer_irq);
+        timer_irq
     } else {
-        info!("[Timer]: Enable APIC periodic mode.");
-        init_periodic_mode()
+        info!(
+            "[Timer]: Enable APIC periodic mode with frequency {} Hz.",
+            TIMER_FREQ
+        );
+        calc_apic_freq_from_pit();
+
+        let timer_irq = IrqLine::alloc().unwrap();
+        init_periodic_mode(&timer_irq);
+        timer_irq
     }
 }
 
-pub(super) static APIC_TIMER_CALLBACK: Once<Arc<dyn Fn() + Sync + Send>> = Once::new();
+pub(super) fn init_ap(timer_irq: &IrqLine) {
+    if is_tsc_deadline_mode_supported() {
+        init_tsc_mode(timer_irq);
+    } else {
+        init_periodic_mode(timer_irq);
+    }
+}
 
 /// Determines if the current system supports tsc_deadline mode APIC timer
 fn is_tsc_deadline_mode_supported() -> bool {
+    use x86::cpuid::cpuid;
+
     const TSC_DEADLINE_MODE_SUPPORT: u32 = 1 << 24;
     let cpuid = cpuid!(1);
     (cpuid.ecx & TSC_DEADLINE_MODE_SUPPORT) > 0
 }
 
-fn init_tsc_mode() -> IrqLine {
-    let timer_irq = IrqLine::alloc().unwrap();
+fn init_tsc_mode(timer_irq: &IrqLine) {
     // Enable tsc deadline mode
     apic::with_borrow(|apic| {
         apic.set_lvt_timer(timer_irq.num() as u64 | (1 << 18));
     });
-    let tsc_step = TSC_FREQ.load(Ordering::Relaxed) / TIMER_FREQ;
 
-    let callback = move || unsafe {
-        let tsc_value = _rdtsc();
-        let next_tsc_value = tsc_step + tsc_value;
-        wrmsr(IA32_TSC_DEADLINE, next_tsc_value);
-    };
+    apic_tsc_deadline_mode_callback();
+}
 
-    callback.call(());
-    APIC_TIMER_CALLBACK.call_once(|| Arc::new(callback));
+pub(super) fn apic_tsc_deadline_mode_callback() {
+    use x86::msr::{wrmsr, IA32_TSC_DEADLINE};
 
-    timer_irq
+    let tsc_value = unsafe { _rdtsc() };
+    let tsc_step = TSC_FREQ.load(Ordering::Relaxed) / TIMER_FREQ;
+    let next_tsc_value = tsc_step + tsc_value;
+    unsafe { wrmsr(IA32_TSC_DEADLINE, next_tsc_value) };
 }
 
-fn init_periodic_mode() -> IrqLine {
+fn init_periodic_mode(timer_irq: &IrqLine) {
+    apic::with_borrow(|apic| {
+        apic.set_timer_init_count(INIT_COUNT.load(Ordering::Relaxed));
+        apic.set_lvt_timer(timer_irq.num() as u64 | (1 << 17));
+        apic.set_timer_div_config(DivideConfig::Divide64);
+    });
+}
+
+static INIT_COUNT: AtomicU64 = AtomicU64::new(0);
```

Comment:
The current code makes me feel like we need something like this:
```rust
enum State {
    PeriodicMode {
        tsc_interval: u32,
    },
    IntervalMode {
        init_count: u32,
    },
}

static state: Once<State> = Once::new();
```

This can replace the magic global atomic values like `IN_APIC_TSC_DEADLINE_MODE` and `INIT_COUNT` and end up with better code clarity and structure.

(`IS_FINISHED` may also become something like `state.is_completed()`.)